### DISCUSSION TITLE
When there is nothing to merge - is FINAL query as fast as regular one?

### DISCUSSION THREAD

**Discussion by xaka** [ClickHouse/ClickHouse#43067] [General]
Hello there,

I'm trying to understand the impact of FINAL on performance in a case of ReplacingMergeTree when there is nothing to merge (i.e. a static table or updates are rare).

Thanks.

**Comment by mo-avatar** [comment]
To my knowledge it is not. according to the official documentation:

Queries that use FINAL are executed slightly slower than similar queries that do not, because:

1.Data is merged during query execution.
2.Queries with FINAL read primary key columns in addition to the columns specified in the query.

besides the first issue, the second one still exists. So, it is not likely that the query is as faster as regular one.

**Comment by den-crane** [comment]
Queries with `final`  N times slower than without final  

```sql
create table X ( A Int64, B Int64) Engine=ReplacingMergeTree order by A as select number,number from numbers(1e9);
optimize table X final;

select avg(A) from X where B=42;
Elapsed: 2.278 sec. Processed 1.00 billion rows

select avg(A) from X Final where B=42;
1 row in set. Elapsed: 13.600 sec. Processed 1.00 billion rows

select avg(A) from X Final where B=42 settings do_not_merge_across_partitions_select_final=1;
1 row in set. Elapsed: 4.479 sec. Processed 1.00 billion rows
```

**Comment by clickhouse-robot-curie** [comment]
The setting `do_not_merge_across_partitions_select_final` is also needed.
Comparison:

```
create table X ( A Int64, B Int64) Engine=ReplacingMergeTree order by A as select number,number from numbers(1e9);
optimize table X final;

select avg(A) from X where B=42;
Elapsed: 0.149 sec

select avg(A) from X Final where B=42;
Elapsed: 2.399 sec.

SET do_not_merge_across_partitions_select_final = 1;
select avg(A) from X Final where B=42;
Elapsed: 0.877 sec.

SET do_not_merge_across_partitions_select_final = 1, max_final_threads = 64;
Elapsed: 0.309 sec.
```

**Comment by nickitat** [comment]
if you're able to maintain each partition to be a single part, there should be no overhead from final during query execution.
speaking of the den-crane@ example, the two times difference in query running time is caused by inability to move the condition to `prewhere`. if `B` is included in sorting key, results will be the same:

``` sql
:) select avg(A) from X final where B=42 settings optimize_move_to_prewhere = 1, optimize_move_to_prewhere_if_final = 1

┌─avg(A)─┐
│     42 │
└────────┘

1 row in set. Elapsed: 1.176 sec. Processed 1.00 billion rows, 8.00 GB (850.24 million rows/s., 6.80 GB/s.)

:) select avg(A) from X where B=42 settings optimize_move_to_prewhere = 1, optimize_move_to_prewhere_if_final = 1

┌─avg(A)─┐
│     42 │
└────────┘

1 row in set. Elapsed: 1.180 sec. Processed 1.00 billion rows, 8.00 GB (847.42 million rows/s., 6.78 GB/s.)
```

relevant issue to `prewhere` problem is https://github.com/ClickHouse/ClickHouse/issues/31411
